substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrManager.java

```diff
@@ -217,6 +204,43 @@ private static Long parseDuration(Map<JfrStartArgument, String> args, JfrStartAr
         return null;
     }
 
+    private static Long parseMaxSize(Map<JfrStartArgument, String> args, JfrStartArgument key) {
+        final String value = args.get(key);
+        if (value != null) {
+            try {
+                int idx = indexOfFirstNonDigitCharacter(value);
+                long number;
+                try {
+                    number = Long.parseLong(value.substring(0, idx));
+                } catch (NumberFormatException e) {
+                    throw new IllegalArgumentException("Expected a number.");
+                }
+
+                // Missing unit, number is plain bytes
+                if (idx == value.length()) {
+                    return number;
+                }
+
+                final char unit = value.substring(idx).charAt(0);
+                switch (unit) {
+                    case 'k':
+                    case 'K':
+                        return number * 1024;
+                    case 'm':
+                    case 'M':
+                        return number * 1024 * 1024;
+                    case 'g':
+                    case 'G':
+                        return number * 1024 * 1024 * 1024;
```

Comment:
In following https://github.com/openjdk/jdk/blob/master/src/hotspot/share/services/diagnosticArgument.cpp#L328 I think this just needs an addition to support the default case. (Code is similar in jdk11)

Everything else looks fine!

Reply:
@jiekang  See https://github.com/oracle/graal/pull/3639/files/355161c32f3a1bf3b5e09826527f32503fdf0d8c#r681255719  

Would you like me to refactor that 
```java
                // Missing unit, number is plain bytes
                if (idx == value.length()) {
                    return number;
```

to a `default` switch branch or is it O.K. like this after all?

Reply:
I think there's a slight difference in that the default switch catches multipliers or delimiters not in the given set, treating them as bytes, while the early return catches the no multiplier case.

E.g. it seems with the HotSpot code, something like `maxsize=5c`, would be accepted as a bytes while here we would throw an exception. I'm not sure I agree with that, but for best transfer of existing settings, it would be easiest to support the same.

Reply:
Though let me actually test that to see what happens with an OpenJDK build; will be later this evening though.

Reply:
@jiekang You are right. I've just tested that. It is counter intuitive, but HotSpot really gracefully ignores unknown units and treats the number as bytes:

```

$ java -XX:+FlightRecorder -XX:StartFlightRecording=maxsize=100000c,filename=flight-java.jfr -Xlog:jfr Main 
[0.370s][info][jfr] Started recording "1" (1) {maxsize=97.7kB, dumponexit=true, filename=/tmp/rep/flight-java.jfr}


$ ./main -XX:+FlightRecorder -XX:StartFlightRecording=maxsize=100000c,filename=flight-java.jfr -XX:FlightRecorderLogging=jfr
Exception in thread "main" com.oracle.svm.core.util.UserError$UserException: Could not parse JFR argument 'maxsize=100000c'. Unit is invalid.
    at com.oracle.svm.core.util.UserError.abort(UserError.java:68)
    at com.oracle.svm.jfr.JfrManager.parseMaxSize(JfrManager.java:238)
    at com.oracle.svm.jfr.JfrManager.initRecording(JfrManager.java:107)
    at com.oracle.svm.jfr.JfrManager.setup(JfrManager.java:72)
    at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:125)
    at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:75)
```

Reply:
@jiekang I modified the behavior as  you suggested.  native-image and HotSpot behave the same now:

```
$  for maxsize in 100000c 100000 100000x 100000xxx 100k 100K 10M 10m 1g 1G 1gg; do echo;echo "maxsize: $maxsize";echo -e "Native-image:\c";./main -XX:+FlightRecorder -XX:StartFlightRecording=maxsize=$maxsize -XX:FlightRecorderLogging=jfr 2>&1 | grep "maxsize=";echo -e "HotSpot:\c";java -XX:+FlightRecorder -XX:StartFlightRecording=maxsize=$maxsize -Xlog:jfr 2>&1 | grep "maxsize=";done

maxsize: 100000c
Native-image:[info][jfr] Started recording "1" (1) {maxsize=97.7kB}
HotSpot:[0.334s][info][jfr] Started recording "1" (1) {maxsize=97.7kB}

maxsize: 100000
Native-image:[info][jfr] Started recording "1" (1) {maxsize=97.7kB}
HotSpot:[0.325s][info][jfr] Started recording "1" (1) {maxsize=97.7kB}

maxsize: 100000x
Native-image:[info][jfr] Started recording "1" (1) {maxsize=97.7kB}
HotSpot:[0.345s][info][jfr] Started recording "1" (1) {maxsize=97.7kB}

maxsize: 100000xxx
Native-image:[info][jfr] Started recording "1" (1) {maxsize=97.7kB}
HotSpot:[0.347s][info][jfr] Started recording "1" (1) {maxsize=97.7kB}

maxsize: 100k
Native-image:[info][jfr] Started recording "1" (1) {maxsize=100.0kB}
HotSpot:[0.345s][info][jfr] Started recording "1" (1) {maxsize=100.0kB}

maxsize: 100K
Native-image:[info][jfr] Started recording "1" (1) {maxsize=100.0kB}
HotSpot:[0.342s][info][jfr] Started recording "1" (1) {maxsize=100.0kB}

maxsize: 10M
Native-image:[info][jfr] Started recording "1" (1) {maxsize=10.0MB}
HotSpot:[0.322s][info][jfr] Started recording "1" (1) {maxsize=10.0MB}

maxsize: 10m
Native-image:[info][jfr] Started recording "1" (1) {maxsize=10.0MB}
HotSpot:[0.332s][info][jfr] Started recording "1" (1) {maxsize=10.0MB}

maxsize: 1g
Native-image:[info][jfr] Started recording "1" (1) {maxsize=1.0GB}
HotSpot:[0.334s][info][jfr] Started recording "1" (1) {maxsize=1.0GB}

maxsize: 1G
Native-image:[info][jfr] Started recording "1" (1) {maxsize=1.0GB}
HotSpot:[0.353s][info][jfr] Started recording "1" (1) {maxsize=1.0GB}

maxsize: 1gg
Native-image:[info][jfr] Started recording "1" (1) {maxsize=1.0GB}
HotSpot:[0.324s][info][jfr] Started recording "1" (1) {maxsize=1.0GB}

```

Reply:
Excellent, thank you!